js-packages/@prestojs/ui-antd/src/widgets/FileWidget.tsx

```diff
@@ -1,105 +1,317 @@
-            </Upload>
-        );
-    }
-);
+    return fileList;
+}
 
-export default FileWidget;
+/**
+ * File upload widget that wraps the antd [Upload](https://ant.design/components/upload/#API) component. Unlike
+ * that component this one never uploads immediately - the raw `File` object will be passed to `onChange` and it's
+ * expected the upload will happen externally (eg. on a form submit).
+ *
+ * For very custom requirements consider using the [useFileList](doc:useFileList) in a custom component.
+ *
+ * // TODO: Example: limit
+ * // TODO: Example: multiple
+ * // TODO: Example: custom children
+ * // TODO: Example: drag and drop
+ * // TODO: beforeUpload
+ *
+ * @extract-docs
+ * @menu-group Widgets
+ * @forward-ref
+ */
+function FileWidget(
+    props: UploadWidgetProps<File, HTMLInputElement>,
+    ref: React.RefObject<Upload>
+): React.ReactElement {
+    let {
+        input,
+        children,
+        beforeUpload,
+        listType = 'text',
+        limit = 1,
+        multiple = limit !== 1,
+        ...rest
+    } = props;
+    const { value, onChange } = input;
+    if (Array.isArray(value) && !multiple) {
+        throw new Error(`When 'value' is an array 'multiple' must be set to true`);
+    }
+    if (value && multiple && !Array.isArray(value)) {
+        throw new Error(`When 'multiple' is true 'value' must be an array`);
+    }
+    const beforeUploadFile = (file: File, files: RcFile[]): false | Promise<void> => {
+        if (beforeUpload) {
+            // beforeUpload can:
+            // - return false to stop upload from occurring
+            // - return a new Blob to replace uploaded File (eg. to implement Crop)
+            // - return undefined to do nothing
+            // - Any of the above via a Promise
+            const response = beforeUpload(file as RcFile, files);
+            if (response === false) {
+                return false;
+            }
+            if (isPromise(response)) {
+                return response.then(file => {
+                    if (multiple) {
+                        // We know value is either not set or an array due to the checks above
+                        const currentValue = (value ?? []) as (string | File)[];
+                        onChange([...currentValue, file]);
+                    } else {
+                        onChange(file);
+                    }
+                });
+            }
+        }
+        if (multiple) {
+            // We know value is either not set or an array due to the checks above
+            const currentValue = (value ?? []) as (string | File)[];
+            onChange([...currentValue, file]);
+        } else {
+            onChange(file);
+        }
+        return false;
+    };
+    const handleRemove = (f: FileWidgetUploadFile): void => {
+        if (!multiple) {
+            onChange(null);
+        } else if (value) {
+            onChange((value as (string | File)[]).filter(key => key !== f.key));
+        }
```

Comment:
Can you swap those if clauses: this is different from the other places where `multiple` is checked (the `multiple` part always comes first)

Reply:
If multiple is true then value will always be true (will always be a non-empty array otherwise you can't have clicked on the handleRemove button) in which case the same if pattern as the others applies:

```
        if (multiple) {
            onChange((value as (string | File)[]).filter(key => key !== f.key));
        } else {
            onChange(null);
        }
```

(ie unless I've read this wrong the check for value is unnecessary)

Reply:
yep good point, fixed